employee-rostering-standalone/src/main/java/org/optaweb/employeerostering/standalone/WebConfig.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.employeerostering.standalone;
+
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.servlet.config.annotation.ViewControllerRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+@Configuration
+public class WebConfig implements WebMvcConfigurer {
+
+    @Override
+    public void addViewControllers(ViewControllerRegistry registry) {
+        // Arcane Spring Boot Magic from https://stackoverflow.com/a/42998817
+        // From what I can tell:
+        // registry.addViewController(String) add a new handler for URLs,
+        // and controller.setViewName(String) tell the handler what to do.
+        // In this case, controller.setViewName("forward:/") tell the handler
+        // to redirect any requests it handles to "/" (which is index.html,
+        // where our frontend exists.
+        
+        // probably handles requests with no trailing paths (/app, /admin,
+        // but not /employees/0)
+        registry.addViewController("/{spring:\\w+}") 
+              .setViewName("forward:/");
+        
+        // probably handles requests with trailing paths (/employees/0 but
+        // not /app)
+        registry.addViewController("/**/{spring:\\w+}")
+              .setViewName("forward:/");
+        
+        // I have no idea what this does; does react have javascript files
+        // with a $ suffix?
+        registry.addViewController("/{spring:\\w+}/**{spring:?!(\\.js|\\.css)$}") 
+              .setViewName("forward:/");
```

Comment:
I guess `$` here is not literal `$` but a meta-character that matches the end of the URL, for example `/app/style.css`. Seems suspicious. I don't think we want to forward such requests to `/index.html`.

Reply:
I see "?!" in the beginning of the second part, which could mean "An optional part that does NOT end in .js or .css". So `/app/icon.png` would match but `/app/code.js` would not. I will see what happens if I remove it.

Reply:
Found this: https://docs.spring.io/spring/docs/5.2.2.RELEASE/spring-framework-reference/web.html#mvc-ann-requestmapping-uri-templates.

> Spring MVC uses the PathMatcher contract and the AntPathMatcher implementation from spring-core for URI path matching.

Also true for [ViewControllerRegistry#addViewController()](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/servlet/config/annotation/ViewControllerRegistry.html#addViewController-java.lang.String-).

According to [AntPathMatcher](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/AntPathMatcher.html), the path patterns only allow basic rules (`?`, `*`, `**`) but **path variables are matched using [regexp](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html)**. `(?!X)` means "X, via zero-width negative lookahead". [Negative lookahead](http://blackwasp.co.uk/RegexLookahead.aspx) asserts that the pattern in the lookahead must not follow the text matched by the initial part of the pattern. So `/{spring:\\w+}/**{spring:?!(\\.js|\\.css)$}` matches paths that have at least two elements and the last one doesn't end with `.js` or `.css`.

At least this is my understanding based on the documentation and source code of `AntPathMatcher`. It would be really interesting to see a unit test demonstrating which URIs are actually matched by the last mapping pattern and which aren't. Also, it would be nice to investigate whether we need all of the three mappings or whether it could be simplified.

Reply:
It manage to work with only the middle matcher (`registry.addViewController("/**/{spring:\\w+}")`)